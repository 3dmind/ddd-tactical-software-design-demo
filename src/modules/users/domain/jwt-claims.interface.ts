export interface JwtClaimsInterface {
  username: string;
}
