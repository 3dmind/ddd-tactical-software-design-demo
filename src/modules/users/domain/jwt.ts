export type JwtToken = string;

export type AccessToken = JwtToken;

export type RefreshToken = JwtToken;
