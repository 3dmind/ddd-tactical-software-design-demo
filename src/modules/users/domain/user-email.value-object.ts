import { isEmail } from 'class-validator';
import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface UserEmailProps {
  value: string;
}

export class UserEmailValueObject extends ValueObject<UserEmailProps> {
  private constructor(props: UserEmailProps) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(email: string): Result<UserEmailValueObject> {
    const nullGuardResult = Guard.againstNullOrUndefined(email, 'email');
    if (!nullGuardResult.succeeded) {
      return Result.fail<UserEmailValueObject>(nullGuardResult.message);
    }

    if (!this.isValidEmail(email)) {
      return Result.fail<UserEmailValueObject>('Email address is not valid.');
    }

    return Result.ok<UserEmailValueObject>(
      new UserEmailValueObject({
        value: this.format(email),
      }),
    );
  }

  public static isValidEmail(email: string): boolean {
    return isEmail(email);
  }

  private static format(email: string): string {
    return email.trim().toLowerCase();
  }
}
