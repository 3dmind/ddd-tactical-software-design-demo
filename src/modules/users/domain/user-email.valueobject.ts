import { isEmail } from 'class-validator';
import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface UserEmailProps {
  value: string;
}

export class UserEmail extends ValueObject<UserEmailProps> {
  private constructor(props: UserEmailProps) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(email: string): Result<UserEmail> {
    const nullGuardResult = Guard.againstNullOrUndefined(email, 'email');
    if (!nullGuardResult.succeeded) {
      return Result.fail<UserEmail>(nullGuardResult.message);
    }

    if (!this.isValidEmail(email)) {
      return Result.fail<UserEmail>('Email address is not valid.');
    }

    return Result.ok<UserEmail>(
      new UserEmail({
        value: this.format(email),
      }),
    );
  }

  public static isValidEmail(email: string): boolean {
    return isEmail(email);
  }

  private static format(email: string): string {
    return email.trim().toLowerCase();
  }
}
