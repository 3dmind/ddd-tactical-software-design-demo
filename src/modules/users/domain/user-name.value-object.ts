import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface UserNamePropsInterface {
  value: string;
}

export class UserNameValueObject extends ValueObject<UserNamePropsInterface> {
  private constructor(props: UserNamePropsInterface) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(username: string): Result<UserNameValueObject> {
    const nullGuardResult = Guard.againstNullOrUndefined(username, 'username');
    if (!nullGuardResult.succeeded) {
      return Result.fail<UserNameValueObject>(nullGuardResult.message);
    }

    const emptyGuardResult = Guard.againstEmpty(username);
    if (!emptyGuardResult.succeeded) {
      return Result.fail<UserNameValueObject>(emptyGuardResult.message);
    }

    return Result.ok<UserNameValueObject>(
      new UserNameValueObject({ value: username }),
    );
  }
}
