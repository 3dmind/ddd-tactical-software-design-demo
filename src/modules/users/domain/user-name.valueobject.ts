import { Guard, Result } from '../../../shared/core';
import { ValueObject } from '../../../shared/domain';

interface UserNameProps {
  value: string;
}

export class UserName extends ValueObject<UserNameProps> {
  private constructor(props: UserNameProps) {
    super(props);
  }

  get value(): string {
    return this.props.value;
  }

  public static create(userName: string): Result<UserName> {
    const nullGuardResult = Guard.againstNullOrUndefined(userName, 'userName');
    if (!nullGuardResult.succeeded) {
      return Result.fail<UserName>(nullGuardResult.message);
    }

    const emptyGuardResult = Guard.againstEmpty(userName);
    if (!emptyGuardResult.succeeded) {
      return Result.fail<UserName>(emptyGuardResult.message);
    }

    return Result.ok<UserName>(new UserName({ value: userName }));
  }
}
