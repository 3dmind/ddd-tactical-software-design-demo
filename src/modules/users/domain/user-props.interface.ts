import { AccessToken, RefreshToken } from './jwt';
import { UserEmailValueObject } from './user-email.value-object';
import { UserNameValueObject } from './user-name.value-object';
import { UserPasswordValueObject } from './user-password.value-object';

export interface UserPropsInterface {
  accessToken?: AccessToken;
  createdAt?: Date;
  email: UserEmailValueObject;
  isEmailVerified?: boolean;
  password: UserPasswordValueObject;
  refreshToken?: RefreshToken;
  username: UserNameValueObject;
}
