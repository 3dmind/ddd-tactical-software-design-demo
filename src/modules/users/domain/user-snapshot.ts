import { AccessToken, RefreshToken } from './jwt';
import { UserEmailValueObject } from './user-email.value-object';
import { UserIdEntity } from './user-id.entity';
import { UserNameValueObject } from './user-name.value-object';
import { UserPasswordValueObject } from './user-password.value-object';
import { UserPropsInterface } from './user-props.interface';

export class UserSnapshot {
  readonly accessToken: AccessToken;
  readonly createdAt: Date;
  readonly email: UserEmailValueObject;
  readonly isEmailVerified: boolean;
  readonly password: UserPasswordValueObject;
  readonly refreshToken: RefreshToken;
  readonly userId: UserIdEntity;
  readonly username: UserNameValueObject;

  constructor(props: Readonly<UserPropsInterface>, userId: UserIdEntity) {
    this.accessToken = props.accessToken;
    this.createdAt = props.createdAt;
    this.email = props.email;
    this.isEmailVerified = props.isEmailVerified;
    this.password = props.password;
    this.refreshToken = props.refreshToken;
    this.userId = userId;
    this.username = props.username;

    Object.freeze(this);
  }
}
