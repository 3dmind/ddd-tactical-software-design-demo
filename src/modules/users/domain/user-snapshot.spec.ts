import { UniqueEntityId } from '../../../shared/domain';
import { UserEmailValueObject } from './user-email.value-object';
import { UserNameValueObject } from './user-name.value-object';
import { UserPasswordValueObject } from './user-password.value-object';
import { UserSnapshot } from './user-snapshot';
import { UserEntity } from './user.entity';

describe('UserSnapshot', () => {
  it('should be defined', () => {
    expect.assertions(2);
    const id = new UniqueEntityId('8597ccd9-4237-44a6-b434-8836693c4b51');
    const username = UserNameValueObject.create('TomTest').getValue();
    const password = UserPasswordValueObject.create({
      value: '123456',
    }).getValue();
    const email = UserEmailValueObject.create('tom.test@gmail.com').getValue();
    const createdAt = new Date(Date.parse('1977-01-01'));
    const userEntity = UserEntity.create(
      {
        createdAt,
        email,
        isEmailVerified: false,
        password,
        username,
      },
      id,
    ).getValue();

    const userSnapshot = new UserSnapshot(userEntity.props, userEntity.userId);

    expect(userSnapshot).toMatchSnapshot();
    expect(Object.isFrozen(userSnapshot)).toBe(true);
  });
});
