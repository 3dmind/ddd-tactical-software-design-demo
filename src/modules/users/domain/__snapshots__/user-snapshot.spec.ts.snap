// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSnapshot should be defined 1`] = `
UserSnapshot {
  "accessToken": "8597ccd9423744a6b4348836693c4b51",
  "createdAt": 1977-01-01T00:00:00.000Z,
  "email": UserEmailValueObject {
    "props": Object {
      "value": "tom.test@gmail.com",
    },
  },
  "isEmailVerified": false,
  "password": UserPasswordValueObject {
    "props": Object {
      "hashed": false,
      "value": "123456",
    },
  },
  "refreshToken": "8597ccd9423744a6b4348836693c4b51",
  "userId": UserIdEntity {
    "_id": UniqueEntityId {
      "value": "8597ccd9-4237-44a6-b434-8836693c4b51",
    },
    "props": null,
  },
  "username": UserNameValueObject {
    "props": Object {
      "value": "TomTest",
    },
  },
}
`;
